packages/vite/src/node/config.ts

```diff
@@ -342,8 +359,10 @@ export async function resolveConfig(
       return p.apply === command
     }
   }) as Plugin[]
-  const [prePlugins, normalPlugins, postPlugins] =
-    sortUserPlugins(rawUserPlugins)
+
+  const [prePlugins, normalPlugins, postPlugins] = sortUserPlugins(
+    rawUserPlugins.concat(command === 'build' ? [] : rawWorkerUserPlugins)
```

Comment:
Would it be confusing for worker plugins be applied to non worker environments in dev?

Reply:
I have tried to separate the worker and user plugins in the development environment, but there is no way to distinguish whether the requested file is a request issued by the worker, so I directly apply the worker plugin to the dev environment. Can it be distinguished by `moduleGraph`?

Reply:
I found [`Sec-Fetch-Dest`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Sec-Fetch-Dest) and [`Referer`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referer) could be used to distinguish the context. Though `Sec-Fetch-Dest` does not work in Safari and `Referer` might be omitted by browser...

|path| `Sec-Fetch-Dest` | `Referer` | description |
|---|----|---|---|
| / | document | http://localhost:3000/ | index.html |
| /@vite/client | script | http://localhost:3000/ | @vite/client |
| /main.js | script | http://localhost:3000/ | script entry |
| /node_modules/vite/dist/client/env.mjs | script | http://localhost:3000/@vite/client | imported from @vite/client |
| /worker.js?worker | script | http://localhost:3000/main.js | content: `export default function WorkerWrapper() { new Worker(...) }` |
| /worker.js?worker_file | worker | http://localhost:3000/ | worker entry |
| /node_modules/vite/dist/client/env.mjs | worker | http://localhost:3000/worker.js?worker_file | imported from worker entry |
| /foo.js | worker | http://localhost:3000/worker.js?worker_file | imported from worker entry |

This data could be obtained by this plugin:
```js
    {
      name: 'a',
      configureServer(s) {
        s.middlewares.use((req, res, next) => {
          console.log(
            req.url,
            req.headers['sec-fetch-dest'],
            req.headers['referer']
          )
          next()
        })
      }
    }
```

Reply:
incomplete realization https://github.com/vitejs/vite/pull/8740/files

Reply:
The referrer trick is interesting. It might be good enough to simplify the plugin flow. The rollup cache findings are neat too, I looks like something we should implement from the get go, given we run `rollup.build` in dev. It seems a bit separate from this PR though but great to have indeed.